db/view/view.cc

```diff
@@ -701,6 +711,14 @@ view_updates::get_view_rows(const partition_key& base_key, const clustering_or_s
         clustering_key ckey = clustering_key::from_range(boost::adaptors::transform(ck, view_managed_key_view_and_action::get_key_view));
         auto action = (action_column < pk.size() ? pk[action_column] : ck[action_column - pk.size()])._action;
         mutation_partition& partition = partition_for(std::move(pkey));
+
+        // Skip adding the row if we already wrote a partition tombstone for this partition, and the update
+        // is deleting the row with an equal row tombstone. This means the entire partition is deleted
+        // so we don't need to generate updates for individual rows.
+        if (partition.partition_tombstone() && update_tomb && update_tomb->tomb() == partition.partition_tombstone()) {
+            return;
```

Comment:
Please correct me if I'm wrong, but it seems to me that if we have the base and view partition key the same, and the update is a partition tombstone, then instead of just writing a view partition tombstone and BE DONE WITH IT, if I understand correctly, we actually read the base partitition - possible a million rows, iterator over them, and just for each of those we avoid writing any update.
If I'm wrong, please tell me that I am.
If I'm not wrong, isn't this a waste? I mean, it's good we don't do a lot of writes, but maybe we also shouldn't do all those reads?

Reply:
Good observation. However, we must consider that the tombstone might have a lower timestamp than some data in the partition. The partition tombstone will only delete data with lower (or equal? not sure) timestamps than the tombstone's timestamp. I don't think we can do the skip you propose, in general.

Reply:
I can probably skip reading the rows as a special case when the optimization is applied on all the views.
In general we work on all the table's views, some may apply the optimization and some not because they have a different partition key.

Reply:
I added a test
Although, I didn't understand why we need to read the rows in this scenario
I think there could be other problems with skipping the read, like if it's a batch update or the optimization doesn't apply to all the views. But maybe it could be skipped in some special cases

Reply:
Now that I thought about it a bit more, I think what I wrote in my previous comment was wrong.

First and foremost, in the case that interests us (i.e. primary key is the same/permuted in both base table and view) and when there are some clustering rows in the base table, we do read the whole partition - see `calculate_affected_clustering_ranges`.

Second, in the case that interests us there are no regular/static columns in the MV partition key, so we do not need to "move" rows in the materialized view by deleting and re-creating.

Therefore, it looks like we can indeed skip reading the partition if we only have views of the type that interests us in this PR. The partition tombstone will cover rows/cells with lower timestamps than the tombstone's timestamp and will keep the ones with higher timestamps. I don't think there is a need to generate updates for the rows/cells/etc. which have higher timestamps - they will just not be covered by the tombstones.

Sorry for the confusion. I think this is a worthwhile optimization to add.

I hope that I didn't miss anything in my analysis this time, @nyh - maybe you saw something in my original comment that I am missing right now and and my previous assessment was correct after all?

Reply:
I added a new commit where I skip iterating over all the rows if the partition tombstone is applied on all the views and there are no other updates